libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+
+	return false;
+}
+
+bool ControlFlowParser::visit(ForStatement const& _forStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	if (auto initializationExpression = _forStatement.initializationExpression())
+		initializationExpression->accept(*this);
+
+	auto condition = createLabelHere();
+
+	if (auto conditionExpression = _forStatement.condition())
+		appendControlFlow(*conditionExpression);
+
+	auto loopExpression = newLabel();
+	auto nodes = splitFlow<2>();
+	auto& afterFor = nodes[1];
+	m_currentNode = nodes[0];
+
+	{
+		BreakContinueScope scope(*this, afterFor, loopExpression);
+		appendControlFlow(_forStatement.body());
+	}
+
+	placeAndConnectLabel(loopExpression);
+
+	if (auto expression = _forStatement.loopExpression())
+		appendControlFlow(*expression);
+
+	connect(m_currentNode, condition);
+	m_currentNode = afterFor;
+
+	return false;
+}
+
+bool ControlFlowParser::visit(WhileStatement const& _whileStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	if (_whileStatement.isDoWhile())
+	{
+		auto afterWhile = newLabel();
+		auto whileBody = createLabelHere();
+
+		{
+			BreakContinueScope scope(*this, afterWhile, whileBody);
+			appendControlFlow(_whileStatement.body());
+		}
+		appendControlFlow(_whileStatement.condition());
+
+		connect(m_currentNode, whileBody);
+		placeAndConnectLabel(afterWhile);
+	}
+	else
+	{
+		auto whileCondition = createLabelHere();
+
+		appendControlFlow(_whileStatement.condition());
+
+		auto nodes = splitFlow<2>();
+
+		auto& whileBody = nodes[0];
+		auto& afterWhile = nodes[1];
+
+		m_currentNode = whileBody;
+		{
+			BreakContinueScope scope(*this, afterWhile, whileCondition);
+			appendControlFlow(_whileStatement.body());
+		}
+
+		connect(m_currentNode, whileCondition);
+
+		m_currentNode = afterWhile;
+	}
+
+
+	return false;
+}
+
+bool ControlFlowParser::visit(Break const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_breakJump, "");
+	connect(m_currentNode, m_breakJump);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Continue const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_continueJump, "");
+	connect(m_currentNode, m_continueJump);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Throw const&)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow->revert, "");
+	connect(m_currentNode, m_cfg.m_currentFunctionFlow->revert);
+	m_currentNode = newLabel();
+	return false;
+}
+
+bool ControlFlowParser::visit(Block const&)
+{
+	solAssert(!!m_currentNode, "");
+	createLabelHere();
+	return true;
+}
+
+void ControlFlowParser::endVisit(Block const&)
+{
+	solAssert(!!m_currentNode, "");
+	createLabelHere();
+}
+
+bool ControlFlowParser::visit(Return const& _return)
+{
+	solAssert(!!m_currentNode, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow, "");
+	solAssert(!!m_cfg.m_currentFunctionFlow->exit, "");
+	// only the first return statement is interesting
```

Comment:
I do not understand this comment.

Reply:
Shouldn't there be an assert if there are multiple?

Reply:
That's true - two consecutive ``return``'s in the same scope used to end up in the same block, but I later changed that, so that after the first return a new disconnected block is started, in which the second return ends up with, which makes much more sense, but I forgot to remove the comment, sorry about that.